### DISCUSSION TITLE
`Column(...,autoincrement=True)` ignored in SQLite dialect

### DISCUSSION THREAD

**Discussion by sizmailov** [sqlalchemy/sqlalchemy#11017] [Usage Questions] [answered]
`Column(...,autoincrement=True)` seems to only trigger an error, but not to emit `AUTOINCREMENT` which is controlled by table's `dialect_options`.

This behavior is at least confusing.

https://github.com/sqlalchemy/sqlalchemy/blob/9b516a9a5fc9e2a1522a5af972bedaeb8926f35b/lib/sqlalchemy/dialects/sqlite/base.py#L1554-L1568

Shouldn't the last condition account for `column.autoincrement`?

```python
if (
      (column.autoincrement or 
      column.table.dialect_options["sqlite"]["autoincrement"])
      and len(column.table.primary_key.columns) == 1
      and issubclass(column.type._type_affinity, sqltypes.Integer)
      and not column.foreign_keys
  ):        
```

**Comment by CaselIT** [comment]
Hi,

The rationale of the current behaviour is documented here https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#sqlite-auto-incrementing-behavior

Also see the linked sqlite documentation that's the basis of the current behaviour https://sqlite.org/autoinc.html

**Comment by zzzeek** [comment, marked as answer]
> `Column(...,autoincrement=True)` seems to only trigger an error, but not to emit `AUTOINCREMENT` which is controlled by table's `dialect_options`.
> 
> This behavior is at least confusing.

it is confusing, and we have a great reason for that - **because sqlite itself is confusing**.     

As discussed in the [linked](https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#sqlite-auto-incrementing-behavior) [documents](https://sqlite.org/autoinc.html), sqlite has **two totally different forms of autoincrement**.   The first one is **implicit**, when you make a column with ``INTEGER PRIMARY KEY``.  **this is the autoincrement feature that should be used with SQLite and it is the one that their own documentation recommends**.   The second one is **explicit**, and changes the way the ``INTEGER PRIMARY KEY`` column works to use some kind of internal convention that quote causes "a slightly different ROWID selection algorithm is used".   **sqlite's documentation explicitly discourages using this feature**.

So if the fact that "autoincrement" on the SQLAlchemy side points to SQLite's **well used, recommended and default mode of integer generation** and not the **never used, explicitly discouraged by SQLite's developers but happens to use the word "autoincrement"**, it should be clear where this problem lies and that is with **sqlite itself, which should rename this keyword to be something like USE_ALTERNATE_ROWID or something not misleading**.      I would recommend raising this issue with them!
